microsoft/testsuites/kselftest/kselftest.py

```diff
@@ -0,0 +1,241 @@
+import os
+import re
+from dataclasses import dataclass
+from pathlib import PurePath, PurePosixPath
+from typing import Any, Dict, List
+
+from assertpy import assert_that
+
+from lisa import Environment, notifier
+from lisa.base_tools.uname import Uname
+from lisa.executable import Tool
+from lisa.messages import SubTestMessage, TestStatus, create_test_result_message
+from lisa.node import Node
+from lisa.operating_system import CBLMariner, Ubuntu
+from lisa.testsuite import TestResult
+from lisa.tools.git import Git
+from lisa.tools.remote_copy import RemoteCopy
+from lisa.tools.tar import Tar
+from lisa.util import LisaException, find_patterns_in_lines
+
+
+@dataclass
+class KselftestResult:
+    name: str = ""
+    status: TestStatus = TestStatus.QUEUED
+    exit_value: int = 0
+
+
+class Kselftest(Tool):
+    repo = "git://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git"
+    mariner_repo = "https://github.com/microsoft/CBL-Mariner-Linux-Kernel"
+    # kselftest result log has "ok" and "not ok" prefixes, use regex to filter them
+    # example: not ok 1 selftests: breakpoints: step_after_suspend_test # exit=126
+    _RESULT_KSELFTEST_OK_REGEX = re.compile(r"ok (.*)\s+")
+
+    @property
+    def command(self) -> str:
+        return str(self._command)
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    def __init__(self, node: Node, kselftest_file_path: str, *args: Any, **kwargs: Any) -> None:
+        super().__init__(node, *args, **kwargs)
+        # command to run kselftests
+        self._command = (
+            self.get_tool_path(use_global=True)
+            / "kselftest-packages/run_kselftest.sh"
+        )
+
+        # tar file path specified in yml
+        self._tar_file_path = kselftest_file_path
+
+        if not self._tar_file_path:
+            # clone kernel, build kernel, then build kselftests
+            uname = self.node.tools[Uname]
+            uname_result = uname.get_linux_information(force_run=False)
+            version=uname_result.kernel_version
+            git = self.node.tools[Git]
+
+            if (version.patch == 0):
+                # clone major.minor kernel from upstream
+                branch = "linux-"
+                branch += str(version.major) + "." + str(version.minor) + ".y"
```

Comment:
You can use f-string, which will be more cleaner and you won't have to use `str`